include/revng/Yield/Support/GraphLayout/SugiyamaStyle/InternalGraph.h

```diff
@@ -0,0 +1,234 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <unordered_map>
+#include <variant>
+
+#include "revng/ADT/GenericGraph.h"
+#include "revng/Yield/Support/GraphLayout/Traits.h"
+
+namespace yield::layout::sugiyama {
+
+class InternalGraph;
+
+namespace detail {
+
+struct InternalNode {
+  layout::Point Center;
+  layout::Size Size;
+
+  bool IsVirtual;
+
+private:
+  std::size_t Index;
+
+private:
+  friend InternalGraph;
+  InternalNode(std::size_t Index, layout::Size &&Size) :
+    Center({ 0, 0 }), Size(std::move(Size)), IsVirtual(false), Index(Index) {}
+  InternalNode(std::size_t Index) :
+    Center({ 0, 0 }), Size({ 0, 0 }), IsVirtual(true), Index(Index) {}
+
+public:
+  InternalNode(const InternalNode &) = delete;
+  InternalNode(InternalNode &&) = default;
+  InternalNode &operator=(InternalNode &&) = default;
+
+  std::size_t index() const { return Index; }
+};
+
+struct InternalEdge {
+  bool IsRouted;
```

Comment:
Can't this be a method that checks if the `Path` is empty?

Reply:
(first see my comment here for additional context: https://github.com/revng/revng/pull/278#discussion_r1149204523)

> What does it even mean to have a not-yet-routed edge with a non-empty path?

Edges can share paths. There is exactly one "`PathOwner`" edge per each real edge in the original graph.

For example, imagine that we have a long edge `A -> B` where rank of `A` is `0`, and rank of `B` is `4`, we need to insert 3 additional nodes in-between, something like
```
A (0) -> A' (1) -> A'' (2) -> A''' (3) -> B (4)
```
Except in the real graph we still only have a single edge for all of these, so only one `LayoutableEdgeTraits::setEdgePath()` call is expected.

So, the approach I chose is to let only one of these edges to own the path (the last one to be specific (`A''' (3) -> B (4)`)), while all the other just refer to it.

By the way the layouter is structured, we are guaranteed that all these edges are going to be routed in the correct order (that's what `orderEdges` (see `EdgeRouting.cpp`) is for), but we still have the problem of making sure we every real edge exactly once. Which is easy to achieve by guaranteeing that only "`PathOwner`"s are exported.

> What happens then when the non-routed-edge-with-non-empty-path gets routed?

By definition, when an edge is routed it simply appends points to the path it's set to. Since we take extra steps to guarantee the proper edge routing order, this let's us route edges without regard to where it's actually stored - by only exposing `appendPoint` which does "the right thing" (tm).

>  

P. S. As for the `IsRouted` flag, we only use to make sure no edge is routed twice, which is significant for "pre-routed" backedges (which we have to route separately because to get them right we need to look at multiple edges at once).

Reply:
P. P. S. I'm guessing I should in fact fit this comment somewhere into the sources as well. What do you think?

Reply:
Ok, this now makes sense, but I have to say it seems more unintuitive and convoluted than necessary. Is this also the only case why you have the `variant` in the `Path`?

Overall I think that if we can remove this asymmetry and just let all edges own their path (allowing duplicated paths of course for all the edges that "share" the path in the current implementation), everything becomes clearer and more intuitive.

The only thing that would be left over would be the asymmetry in how you apply these things back to the output graph, but that's an intrinsic asymmetry we'll never get rid of (there's a many-to-one mapping from internal graph to output graph, we can't do anything about it).

> By definition, when an edge is routed it simply appends points to the path it's set to. Since we take extra steps to guarantee the proper edge routing order, this let's us route edges without regard to where it's actually stored - by only exposing appendPoint which does "the right thing" (tm).

:thinking: this makes me think that the routing of multi-node-edges happens iteratively and not all at once. Is that the case? Would that prevent or complicate the implementation of what I just proposed above in this message?

> P. S. As for the IsRouted flag, we only use to make sure no edge is routed twice, which is significant for "pre-routed" backedges (which we have to route separately because to get them right we need to look at multiple edges at once).

I see, but this can also be done in my proposal by just pre-filling the `Path` for pre-routed edges. So that if a edge has non-empty path it means that is routed already (possibly pre-routed) and we don't route it again.

> P. P. S. I'm guessing I should in fact fit this comment somewhere into the sources as well. What do you think? Any ideas for an optimal location?

Yes, all this information needs to be put in some comments. Let's just finalize the design, and then we can come up with where's the best place to document this.

Reply:
Anyway, independently of how this discussion turns out, the proper place to document this is on the `variant`. Unless we manage to get rid of the `variant`, but at that point also the `IsRouted` thing and all this discussion is outdated. Which BTW seems like a good reason to get rid of the `variant` if we can, because all this non-obvious information becomes irrelevant.

Reply:
I'm closing this as per the mattermost discussion.